lib/elixir/lib/calendar/iso.ex

```diff
@@ -337,6 +342,122 @@ defmodule Calendar.ISO do
     Integer.mod(date_to_iso_days(year, month, day) + 5, 7) + 1
   end
 
+  @doc """
+  Calculates the day of the year from the given `year`, `month`, and `day`.
+
+  It is an integer from 1 to 366.
+
+  ## Examples
+
+      iex> Calendar.ISO.day_of_year(2016, 1, 31)
+      31
+      iex> Calendar.ISO.day_of_year(-99, 2, 1)
+      32
+      iex> Calendar.ISO.day_of_year(2018, 2, 28)
+      59
+
+  """
+  @doc since: "1.8.0"
+  @spec day_of_year(year, month, day) :: 1..366
+  @impl true
+  def day_of_year(year, month, day)
+      when is_integer(year) and is_integer(month) and is_integer(day) do
+    date_to_iso_days(year, month, day) - date_to_iso_days(year, 1, 1) + 1
+  end
+
+  @doc """
+  Calculates the quarter of the year from the given `year`, `month`, and `day`.
+
+  It is an integer from 1 to 4.
+
+  ## Examples
+
+      iex> Calendar.ISO.quarter_of_year(2016, 1, 31)
+      1
+      iex> Calendar.ISO.quarter_of_year(2016, 4, 3)
+      2
+      iex> Calendar.ISO.quarter_of_year(-99, 9, 31)
+      3
+      iex> Calendar.ISO.quarter_of_year(2018, 12, 28)
+      4
+
+  """
+  @doc since: "1.8.0"
+  @spec quarter_of_year(year, month, day) :: 1..4
+  @impl true
+  def quarter_of_year(year, month, day)
+      when is_integer(year) and is_integer(month) and is_integer(day) do
+    cond do
+      month <= 3 -> 1
+      month <= 6 -> 2
+      month <= 9 -> 3
+      true -> 4
+    end
+  end
+
+  @doc """
+  Calculates the year and era from the given `year`.
+
+  The ISO calendar has two eras: the current era which
+  starts in year 1 and is defined as era "1". And a
+  second era for those years less than 1 defined as
+  era "0".
+
+  ## Examples
+
+      iex> Calendar.ISO.year_of_era(1)
+      {1, 1}
+      iex> Calendar.ISO.year_of_era(2018)
+      {2018, 1}
+      iex> Calendar.ISO.year_of_era(0)
+      {1, 0}
+      iex> Calendar.ISO.year_of_era(-1)
+      {2, 0}
+
+  """
+  @doc since: "1.8.0"
+  @spec year_of_era(year) :: {year, era :: 0..1}
+  @impl true
+  def year_of_era(year) when is_integer(year) and year > 0 do
+    {year, 1}
+  end
+
+  def year_of_era(year) when is_integer(year) and year < 1 do
```

Comment:
We could probably leave out the year check in the other clause since it will always be true. Unfortunately, currently the compiler is not able to make that deduction, so it leaves in the check in the final code.

Reply:
This check will hardly matter though in terms of performance though. We can leave it, it is fine.